Massive PR - Refactored text rendering - alignment, word wrap, & hotkeys all now work with Unicode

tig opened the pull request:
This PR fixes a whole swath of known bugs (and many lurking) in how text is handled throughout the library. I was frustrated playing whack-a-mole on layout and formatting issues when dealing with edge cases (especially with Unicode); I thought hard about how to reduce the amount of text processing code in the project; to centralize the most complex and error prone into one place. This way bugs found get fixed once, and the fix goes everywhere.

No changes to demo.cs were required after doing all this; e.g. no breaking changes.

Below is the essence of these changes:

      * View now has a Text property, implemented via the new TextFormatting class. (Thanks @tig!)
        * TextAlignment is implemented once across all Views that support it.
        * Unicode support is now much more robust and complete; dozens of bugs fixed.
        * Any view derived from View now has a Text property with multi-line text formatting, including word-wrap and hotkey support.
        * Any view derived from View now gets mouse click (Clicked event) support for free.
        * Label is now just an alias for View.
        * Button is now a very thin class derived from View (no API changes).
        * Dozens of unit tests for TextAlignment are provided reducing the chance of regressions.
     * I found and fixed unicode related bugs in almost every other class. 


Closes:
* Closes #464
* Closes #459
* Closes #353
* Closes #326
* Closes #182
* Enables progress on #134
* Closes #36 
* Closes #31 
* Further progress on #508

BDisp commented:
Very good. In the Time and Date scenario it is not capturing the event when the time or date is changed.

tig commented:
I just updated this PR with a merge of @bdisp's fix of #705 and the change making `LayoutSubviews` be not `virtual`.

tig commented:
FWIW, I'm still waiting for @migueldeicaza to approve this PR.

migueldeicaza commented:
Just got the ping on the new GitHub app - I will review tonight, apologies for the delay!

BDisp commented:
@tig, `Label`'s views does not update the text change. It's needed to changing something in the Time and Date scenario?

migueldeicaza commented on Terminal.Gui/Core/View.cs:
```diff
@@ -167,6 +149,27 @@ internal enum Direction {
+		/// <summary>
+		/// Gets or sets the HotKey defined for this view. A user pressing HotKey on the keyboard while this view has focus will cause the Clicked event to fire.
+		/// </summary>
+		public Key HotKey { get => viewText.HotKey; set => viewText.HotKey = value; }
```
Is there a way for a View to not have a hotkey specified?  I am thinking about entry lines, or lists, not sure how this is handled in this new setup.

tig replied:
Yes, by default the hotkey capability is 'disabled' because the `HotKeySpecifier` defaults to 0xffff.

migueldeicaza replied:
Thanks!

migueldeicaza commented on Terminal.Gui/Core/View.cs:
```diff
@@ -167,6 +149,27 @@ internal enum Direction {
+		///   raised when the button is activated either with
+		///   the mouse or the keyboard.
+		/// </remarks>
+		public Action Clicked;
```
I am not sure why Clicked needs to be moved to the View class, Clicked is really a behavior that makes sense for some classes, not all of them.

tig commented:
Great comments. I appreciate the thoughtfulness.

Go look at Button and how it leverages the base Text property. I think it's pretty elegant. The idea is a subclass can very easily leverage the hotkey stuff in View and still expose Text where callers just get back what they set.

-cek
Sent from mobile device. Expect brevity, typos, and dangerous driving.

________________________________
From: Miguel de Icaza <notifications@github.com>
Sent: Saturday, June 20, 2020 8:14:05 PM
To: migueldeicaza/gui.cs <gui.cs@noreply.github.com>
Cc: Charlie Kindel (charlie@kindel.com) <charlie@kindel.com>; Mention <mention@noreply.github.com>
Subject: Re: [migueldeicaza/gui.cs] Massive PR - Refactored text rendering - alignment, word wrap, & hotkeys all now work with Unicode (#702)


@migueldeicaza commented on this pull request.

I do like the idea behind this patch, but I have some questions.

The best addition is the general purpose text formatting capability that is now delegated to a secondary class. Then there is the bit that I am not quite comfortable with, which is the introduction of the "text" property as being a capability of the view, and the things that go with it (the hotkey, Clicked handler), but also, this makes some parts of the code rely on a "public new string Text" approach that essentially hide the Text property that was added and now conflicts with the other meanings.

I wonder if we could instead keep the first capability (the unified text formatting, and maybe even have some helper methods for any subclasses that render text), but do not introduce the Text into the base class for view.

I am not pushing back, I am merely trying to explore this.

My initial reaction was that "Text" did not belong in View, but then I thought Text is such a key element of a console toolkit, that perhaps this is right, and this is the basic primitive to build on. But I did not quite like that this meant that hotkeys, clicked made it all the way there. Perhaps it is inevitable, but I did not understand the reason, so I left a couple of comments on those lines.

If we decide to go down this path, may I suggest that rather than "Text" we have "BackingText" as the property name, which would indicate "This is the text that this view renders in the course of regular business", and that would at least not conflict with more precise versions of it.

________________________________

In Terminal.Gui/Core/View.cs<https://github.com/migueldeicaza/gui.cs/pull/702#discussion_r443176819>:

> @@ -167,6 +149,27 @@ internal enum Direction {
                /// </summary>
                public Action<MouseEventArgs> MouseClick;

+               /// <summary>
+               /// Gets or sets the HotKey defined for this view. A user pressing HotKey on the keyboard while this view has focus will cause the Clicked event to fire.
+               /// </summary>
+               public Key HotKey { get => viewText.HotKey; set => viewText.HotKey = value; }


Is there a way for a View to not have a hotkey specified? I am thinking about entry lines, or lists, not sure how this is handled in this new setup.

________________________________

In Terminal.Gui/Core/View.cs<https://github.com/migueldeicaza/gui.cs/pull/702#discussion_r443176847>:

> +
+               /// <summary>
+               /// Gets or sets the specifier character for the hotkey (e.g. '_'). Set to '\xffff' to disable hotkey support for this View instance. The default is '\xffff'.
+               /// </summary>
+               public Rune HotKeySpecifier { get => viewText.HotKeySpecifier; set => viewText.HotKeySpecifier = value; }
+
+               /// <summary>
+               ///   Clicked <see cref="Action"/>, raised when the user clicks the primary mouse button within the Bounds of this <see cref="View"/>
+               ///   or if the user presses the action key while this view is focused. (TODO: IsDefault)
+               /// </summary>
+               /// <remarks>
+               ///   Client code can hook up to this event, it is
+               ///   raised when the button is activated either with
+               ///   the mouse or the keyboard.
+               /// </remarks>
+               public Action Clicked;


I am not sure why Clicked needs to be moved to the View class, Clicked is really a behavior that makes sense for some classes, not all of them.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/migueldeicaza/gui.cs/pull/702#pullrequestreview-434466170>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAEO6CVNZQHFQTPPHFON3XDRXV3H3ANCNFSM4OBLITXQ>.

tig commented:
On clicked: as an app developer, I find it very useful (and natural) that I can get clicked events from any subview. Without this, every subview may do it differently. By putting it in view it reduces API surface area, and thus chances for buy

-cek
Sent from mobile device. Expect brevity, typos, and dangerous driving.
________________________________
From: Miguel de Icaza <notifications@github.com>
Sent: Saturday, June 20, 2020 8:14:05 PM
To: migueldeicaza/gui.cs <gui.cs@noreply.github.com>
Cc: Charlie Kindel (charlie@kindel.com) <charlie@kindel.com>; Mention <mention@noreply.github.com>
Subject: Re: [migueldeicaza/gui.cs] Massive PR - Refactored text rendering - alignment, word wrap, & hotkeys all now work with Unicode (#702)


@migueldeicaza commented on this pull request.

I do like the idea behind this patch, but I have some questions.

The best addition is the general purpose text formatting capability that is now delegated to a secondary class. Then there is the bit that I am not quite comfortable with, which is the introduction of the "text" property as being a capability of the view, and the things that go with it (the hotkey, Clicked handler), but also, this makes some parts of the code rely on a "public new string Text" approach that essentially hide the Text property that was added and now conflicts with the other meanings.

I wonder if we could instead keep the first capability (the unified text formatting, and maybe even have some helper methods for any subclasses that render text), but do not introduce the Text into the base class for view.

I am not pushing back, I am merely trying to explore this.

My initial reaction was that "Text" did not belong in View, but then I thought Text is such a key element of a console toolkit, that perhaps this is right, and this is the basic primitive to build on. But I did not quite like that this meant that hotkeys, clicked made it all the way there. Perhaps it is inevitable, but I did not understand the reason, so I left a couple of comments on those lines.

If we decide to go down this path, may I suggest that rather than "Text" we have "BackingText" as the property name, which would indicate "This is the text that this view renders in the course of regular business", and that would at least not conflict with more precise versions of it.

________________________________

In Terminal.Gui/Core/View.cs<https://github.com/migueldeicaza/gui.cs/pull/702#discussion_r443176819>:

> @@ -167,6 +149,27 @@ internal enum Direction {
                /// </summary>
                public Action<MouseEventArgs> MouseClick;

+               /// <summary>
+               /// Gets or sets the HotKey defined for this view. A user pressing HotKey on the keyboard while this view has focus will cause the Clicked event to fire.
+               /// </summary>
+               public Key HotKey { get => viewText.HotKey; set => viewText.HotKey = value; }


Is there a way for a View to not have a hotkey specified? I am thinking about entry lines, or lists, not sure how this is handled in this new setup.

________________________________

In Terminal.Gui/Core/View.cs<https://github.com/migueldeicaza/gui.cs/pull/702#discussion_r443176847>:

> +
+               /// <summary>
+               /// Gets or sets the specifier character for the hotkey (e.g. '_'). Set to '\xffff' to disable hotkey support for this View instance. The default is '\xffff'.
+               /// </summary>
+               public Rune HotKeySpecifier { get => viewText.HotKeySpecifier; set => viewText.HotKeySpecifier = value; }
+
+               /// <summary>
+               ///   Clicked <see cref="Action"/>, raised when the user clicks the primary mouse button within the Bounds of this <see cref="View"/>
+               ///   or if the user presses the action key while this view is focused. (TODO: IsDefault)
+               /// </summary>
+               /// <remarks>
+               ///   Client code can hook up to this event, it is
+               ///   raised when the button is activated either with
+               ///   the mouse or the keyboard.
+               /// </remarks>
+               public Action Clicked;


I am not sure why Clicked needs to be moved to the View class, Clicked is really a behavior that makes sense for some classes, not all of them.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/migueldeicaza/gui.cs/pull/702#pullrequestreview-434466170>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAEO6CVNZQHFQTPPHFON3XDRXV3H3ANCNFSM4OBLITXQ>.

migueldeicaza commented:
Quick reply (just had this realization overnight, and have not read the other comments yet).

I think I see the point of "Clicked", but I think that there is a better approach for this, and it is to hide this behind the model used by UIGestureRecognizer from iOS (which is what this is based on).

The idea is to essentially make it so that it is easier for consumers of a View to add a gesture recognizer (in our case, clicked, or double click) without having to subclass.    I think we can even make it simpler, given the C# language capabilities, but it would look something like this:

```
var clickRecognizer = new ClickRecognizer (()=>{ print ("Clicked"); });
view.AddRecognizer (clickRecognizer);
```

Will read the comments later today, just posting this before breakfast.

tig commented:
> Quick reply (just had this realization overnight, and have not read the other comments yet).
> 
> I think I see the point of "Clicked", but I think that there is a better approach for this, and it is to hide this behind the model used by UIGestureRecognizer from iOS (which is what this is based on).
> 
> The idea is to essentially make it so that it is easier for consumers of a View to add a gesture recognizer (in our case, clicked, or double click) without having to subclass. I think we can even make it simpler, given the C# language capabilities, but it would look something like this:
> 
> ```
> var clickRecognizer = new ClickRecognizer (()=>{ print ("Clicked"); });
> view.AddRecognizer (clickRecognizer);
> ```
> 
> Will read the comments later today, just posting this before breakfast.

My only issues with the "Recognizer" pattern we'd want to be consistent across the whole project and that's a lot of work as we're trying to converge on 1.0. At this point I'd rather just remove it from View for now, put it back directly on Button, and move on towards 1.0. My primary motivation for this PR was to remove duplicate code from the project, esp. parts that are clearly 'bug farms', like text processing; adding `Clicked` to all Views was just a side benefit.

tig commented:
I've removed the `Clicked` event from `View` based on @migueldeicaza's comments. 

Release notes now say:

      * View now has a Text property, implemented via the new TextFormatting class. (Thanks @tig!)
        * TextAlignment is implemented once across all Views that support it.
        * Unicode support is now much more robust and complete; dozens of bugs fixed.
        * Any view dervied from View now has a Text property with multi-line text formatting, including word-wrap and hotkey support.
        * Label is now mostly just an alias for View; supports Clicked
        * Button is now a very thin class derived from View (no API changes).
        * Dozens of unit tests for TextAlignment are provided reducing the chance of regressions.

I think the last remaining debate is this from @migueldeicaza :

> If we decide to go down this path, may I suggest that rather than "Text" we have "BackingText" as the property name, which would indicate "This is the text that this view renders in the course of regular business", and that would at least not conflict with more precise versions of it.

I think this just makes it more complicated. Very few views have the need for both "Text" and "BackingText". For those where having text is key, I think "Text" makes the most sense.

BDisp commented on Terminal.Gui/Core/TextFormatter.cs:
```diff
@@ -0,0 +1,578 @@
+			for (int line = 0; line < Lines.Count; line++) {
+				if (line > bounds.Height)
+					continue;
+				var runes = lines [line].ToRunes ();
```
lines don't have the actual text value and that why the Label isn't changed correctly in the TimeAndDate.cs.

tig replied:
My apologies for missing this comment. I did not see it. I think I now understand and will work on a fix.

migueldeicaza commented:
Thanks for the Clicked change.

Let us get this merged for now, but would love to keep in the back of our minds that I do note like the overriding via "new Text" in those places and that we should come up with an alternative.
